File: android/src/main/java/com/openauth/googlern/GoogleAuthModule.kt
package com.openauth.googlern

import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.security.SecureRandom
import java.util.*

@ReactModule(name = GoogleAuthModule.NAME)
class GoogleAuthModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var clientId: String? = null
    private var redirectUri: String? = null
    private var scopes: List<String> = listOf("openid", "profile", "email")
    private var pendingPromise: Promise? = null
    private var pendingState: String? = null

    override fun getName(): String = NAME

    /**
     * Extract the client ID number from full format or return as-is if already in short format
     * @param clientId Full format (xxx.apps.googleusercontent.com) or short format (xxx)
     * @return Short format client ID (xxx)
     */
    private fun extractClientIdNumber(clientId: String): String {
        return clientId.replace(".apps.googleusercontent.com", "")
    }

    @ReactMethod
    fun configure(config: ReadableMap, promise: Promise) {
        try {
            clientId = config.getString("clientId")
            if (clientId == null) {
                promise.reject("CONFIG_ERROR", "clientId is required")
                return
            }

            // Extract client ID number (handles both full and short formats)
            val clientIdNumber = extractClientIdNumber(clientId!!)
            redirectUri = if (config.hasKey("redirectUri")) {
                config.getString("redirectUri")
            } else {
                "com.googleusercontent.apps.$clientIdNumber:/oauth2callback"
            }

            if (config.hasKey("scopes")) {
                val scopesArray = config.getArray("scopes")
                scopes = (0 until (scopesArray?.size() ?: 0)).map { 
                    scopesArray?.getString(it) ?: "" 
                }
            }

            promise.resolve(
                Arguments.createMap().apply {
                    putBoolean("success", true)
                }
            )
        } catch (e: Exception) {
            promise.reject("CONFIG_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun signIn(promise: Promise) {
        val localClientId = clientId
        if (localClientId == null) {
            promise.reject("NOT_CONFIGURED", "Google Auth not configured. Call configure() first.")
            return
        }

        pendingPromise = promise
        pendingState = generateRandomString(32)
        val nonce = generateRandomString(32)

        // Extract client ID number for redirect URI
        val clientIdNumber = extractClientIdNumber(localClientId)
        val scope = scopes.joinToString(" ")
        val localRedirectUri = redirectUri ?: "com.googleusercontent.apps.$clientIdNumber:/oauth2callback"

        val authUrl = Uri.parse("https://accounts.google.com/o/oauth2/v2/auth").buildUpon()
            .appendQueryParameter("client_id", localClientId)
            .appendQueryParameter("redirect_uri", localRedirectUri)
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("scope", scope)
            .appendQueryParameter("state", pendingState)
            .appendQueryParameter("nonce", nonce)
            .appendQueryParameter("access_type", "offline")
            .appendQueryParameter("prompt", "consent")
            .build()

        try {
            val activity = reactApplicationContext.currentActivity
            if (activity == null) {
                promise.reject("NO_ACTIVITY", "No current activity")
                return
            }

            val customTabsIntent = CustomTabsIntent.Builder().build()
            customTabsIntent.launchUrl(activity, authUrl)
        } catch (e: Exception) {
            promise.reject("LAUNCH_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun signOut(promise: Promise) {
        try {
            promise.resolve(
                Arguments.createMap().apply {
                    putBoolean("success", true)
                }
            )
        } catch (e: Exception) {
            promise.reject("SIGNOUT_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun getCurrentUser(promise: Promise) {
        // Return null - JS layer handles this with TokenManager
        promise.resolve(null)
    }

    @ReactMethod
    fun handleOpenURL(url: String, promise: Promise) {
        val uri = Uri.parse(url)
        val state = uri.getQueryParameter("state")
        val code = uri.getQueryParameter("code")
        val error = uri.getQueryParameter("error")

        val currentPromise = pendingPromise
        if (currentPromise == null) {
            promise.reject("NO_PENDING_AUTH", "No pending authentication")
            return
        }

        if (state != pendingState) {
            currentPromise.reject("STATE_MISMATCH", "State parameter mismatch")
            pendingPromise = null
            return
        }

        if (error != null) {
            currentPromise.reject("AUTH_ERROR", error)
            pendingPromise = null
            return
        }

        if (code == null) {
            currentPromise.reject("NO_CODE", "No authorization code received")
            pendingPromise = null
            return
        }

        // Exchange code for tokens
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val result = exchangeCodeForTokens(code)
                withContext(Dispatchers.Main) {
                    currentPromise.resolve(result)
                    pendingPromise = null
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    currentPromise.reject("TOKEN_EXCHANGE_ERROR", e.message, e)
                    pendingPromise = null
                }
            }
        }

        promise.resolve(true)
    }

    private suspend fun exchangeCodeForTokens(code: String): WritableMap {
        val localClientId = clientId ?: throw Exception("Client ID not configured")
        // Extract client ID number for redirect URI
        val clientIdNumber = extractClientIdNumber(localClientId)
        val localRedirectUri = redirectUri ?: "com.googleusercontent.apps.$clientIdNumber:/oauth2callback"

        val url = URL("https://oauth2.googleapis.com/token")
        val connection = url.openConnection() as HttpURLConnection
        
        try {
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            connection.doOutput = true

            val postData = "code=$code&client_id=$localClientId&redirect_uri=$localRedirectUri&grant_type=authorization_code"
            
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(postData)
                writer.flush()
            }

            val responseCode = connection.responseCode
            if (responseCode != HttpURLConnection.HTTP_OK) {
                throw Exception("Token exchange failed with code: $responseCode")
            }

            val response = BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
                reader.readText()
            }

            val jsonResponse = JSONObject(response)
            val accessToken = jsonResponse.getString("access_token")

            // Get user info
            val userInfo = getUserInfo(accessToken)

            return Arguments.createMap().apply {
                putString("accessToken", jsonResponse.optString("access_token", ""))
                putString("refreshToken", jsonResponse.optString("refresh_token", ""))
                putInt("expiresIn", jsonResponse.optInt("expires_in", 3600))
                putString("idToken", jsonResponse.optString("id_token", ""))
                putString("tokenType", jsonResponse.optString("token_type", "Bearer"))
                putMap("user", userInfo)
            }
        } finally {
            connection.disconnect()
        }
    }

    private fun getUserInfo(accessToken: String): WritableMap {
        val url = URL("https://www.googleapis.com/oauth2/v2/userinfo")
        val connection = url.openConnection() as HttpURLConnection
        
        try {
            connection.setRequestProperty("Authorization", "Bearer $accessToken")
            
            val response = BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
                reader.readText()
            }

            val json = JSONObject(response)

            return Arguments.createMap().apply {
                putString("id", json.optString("id", ""))
                putString("email", json.optString("email", ""))
                putBoolean("emailVerified", json.optBoolean("verified_email", false))
                putString("name", json.optString("name", ""))
                putString("givenName", json.optString("given_name", ""))
                putString("familyName", json.optString("family_name", ""))
                putString("photoUrl", json.optString("picture", ""))
                putString("locale", json.optString("locale", ""))
            }
        } finally {
            connection.disconnect()
        }
    }

    private fun generateRandomString(length: Int): String {
        val random = SecureRandom()
        val bytes = ByteArray(length)
        random.nextBytes(bytes)
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes).substring(0, length)
    }

    companion object {
        const val NAME = "GoogleAuthModule"
    }
}

File: ios/GoogleAuthModule.swift
//
//  GoogleAuthModule.swift
//  thewoowonGoogleRN
//
//  Created by thewoowon
//  Copyright © 2024. All rights reserved.
//

import Foundation
import AuthenticationServices
import React

@objc(GoogleAuthModule)
class GoogleAuthModule: RCTEventEmitter {

    private var authSession: ASWebAuthenticationSession?
    private var clientId: String?
    private var redirectUri: String?
    private var scopes: [String] = ["openid", "profile", "email"]

    // MARK: - Helper Methods

    /// Extract the client ID number from full format or return as-is if already in short format
    /// - Parameter clientId: Full format (xxx.apps.googleusercontent.com) or short format (xxx)
    /// - Returns: Short format client ID (xxx)
    private func extractClientIdNumber(_ clientId: String) -> String {
        return clientId.replacingOccurrences(of: ".apps.googleusercontent.com", with: "")
    }

    // MARK: - Configuration
    
    @objc
    func configure(_ config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let clientId = config["clientId"] as? String else {
            reject("CONFIG_ERROR", "clientId is required", nil)
            return
        }

        // Extract client ID number (handles both full and short formats)
        let clientIdNumber = extractClientIdNumber(clientId)
        self.clientId = clientId
        self.redirectUri = config["redirectUri"] as? String ?? "com.googleusercontent.apps.\(clientIdNumber):/oauth2callback"

        if let scopesArray = config["scopes"] as? [String] {
            self.scopes = scopesArray
        }

        resolve(["success": true])
    }
    
    // MARK: - Sign In
    
    @objc
    func signIn(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let clientId = self.clientId else {
            reject("NOT_CONFIGURED", "Google Auth not configured. Call configure() first.", nil)
            return
        }

        let state = generateRandomString(length: 32)
        let nonce = generateRandomString(length: 32)

        // Extract client ID number for URL scheme
        let clientIdNumber = extractClientIdNumber(clientId)

        // Build authorization URL
        let scope = scopes.joined(separator: " ")
        let redirectUri = self.redirectUri ?? "com.googleusercontent.apps.\(clientIdNumber):/oauth2callback"

        var components = URLComponents(string: "https://accounts.google.com/o/oauth2/v2/auth")!
        components.queryItems = [
            URLQueryItem(name: "client_id", value: clientId),
            URLQueryItem(name: "redirect_uri", value: redirectUri),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "scope", value: scope),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "nonce", value: nonce),
            URLQueryItem(name: "access_type", value: "offline"),
            URLQueryItem(name: "prompt", value: "consent")
        ]

        guard let authURL = components.url else {
            reject("URL_ERROR", "Failed to create authorization URL", nil)
            return
        }

        DispatchQueue.main.async {
            self.authSession = ASWebAuthenticationSession(
                url: authURL,
                callbackURLScheme: "com.googleusercontent.apps.\(clientIdNumber)"
            ) { [weak self] callbackURL, error in
                guard let self = self else { return }
                
                if let error = error {
                    if (error as NSError).code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
                        reject("CANCELLED", "User cancelled the login", nil)
                    } else {
                        reject("AUTH_ERROR", error.localizedDescription, error)
                    }
                    return
                }
                
                guard let callbackURL = callbackURL else {
                    reject("NO_CALLBACK", "No callback URL received", nil)
                    return
                }
                
                self.handleCallback(callbackURL: callbackURL, state: state, resolve: resolve, reject: reject)
            }
            
            self.authSession?.presentationContextProvider = self
            self.authSession?.prefersEphemeralWebBrowserSession = false
            self.authSession?.start()
        }
    }
    
    // MARK: - Sign Out
    
    @objc
    func signOut(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        // Clear any stored credentials
        // For web-based OAuth, we just acknowledge the sign out
        resolve(["success": true])
    }
    
    // MARK: - Get Current User
    
    @objc
    func getCurrentUser(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        // This would require storing user info locally
        // For now, return null - the JS layer handles this with TokenManager
        resolve(NSNull())
    }
    
    // MARK: - Private Helpers
    
    private func handleCallback(callbackURL: URL, state: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let components = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false),
              let queryItems = components.queryItems else {
            reject("PARSE_ERROR", "Failed to parse callback URL", nil)
            return
        }
        
        var params: [String: String] = [:]
        for item in queryItems {
            if let value = item.value {
                params[item.name] = value
            }
        }
        
        // Verify state
        guard let returnedState = params["state"], returnedState == state else {
            reject("STATE_MISMATCH", "State parameter mismatch", nil)
            return
        }
        
        // Get authorization code
        guard let code = params["code"] else {
            if let error = params["error"] {
                reject("AUTH_ERROR", error, nil)
            } else {
                reject("NO_CODE", "No authorization code received", nil)
            }
            return
        }
        
        // Exchange code for tokens
        exchangeCodeForTokens(code: code, resolve: resolve, reject: reject)
    }
    
    private func exchangeCodeForTokens(code: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let clientId = self.clientId else {
            reject("NOT_CONFIGURED", "Client ID not configured", nil)
            return
        }

        // Extract client ID number for redirect URI
        let clientIdNumber = extractClientIdNumber(clientId)
        let redirectUri = self.redirectUri ?? "com.googleusercontent.apps.\(clientIdNumber):/oauth2callback"
        
        var request = URLRequest(url: URL(string: "https://oauth2.googleapis.com/token")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        let body = [
            "code": code,
            "client_id": clientId,
            "redirect_uri": redirectUri,
            "grant_type": "authorization_code"
        ]
        
        request.httpBody = body.percentEncoded()
        
        let task = URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                reject("NETWORK_ERROR", error.localizedDescription, error)
                return
            }
            
            guard let data = data else {
                reject("NO_DATA", "No data received", nil)
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    if let errorMsg = json["error"] as? String {
                        reject("TOKEN_ERROR", errorMsg, nil)
                        return
                    }
                    
                    self?.getUserInfo(tokenResponse: json, resolve: resolve, reject: reject)
                }
            } catch {
                reject("PARSE_ERROR", "Failed to parse token response", error)
            }
        }
        
        task.resume()
    }
    
    private func getUserInfo(tokenResponse: [String: Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let accessToken = tokenResponse["access_token"] as? String else {
            reject("NO_ACCESS_TOKEN", "No access token in response", nil)
            return
        }
        
        var request = URLRequest(url: URL(string: "https://www.googleapis.com/oauth2/v2/userinfo")!)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                reject("USERINFO_ERROR", error.localizedDescription, error)
                return
            }
            
            guard let data = data else {
                reject("NO_DATA", "No user info data received", nil)
                return
            }
            
            do {
                if let userInfo = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    let result: [String: Any] = [
                        "accessToken": tokenResponse["access_token"] ?? "",
                        "refreshToken": tokenResponse["refresh_token"] ?? "",
                        "expiresIn": tokenResponse["expires_in"] ?? 3600,
                        "idToken": tokenResponse["id_token"] ?? "",
                        "tokenType": tokenResponse["token_type"] ?? "Bearer",
                        "user": [
                            "id": userInfo["id"] ?? "",
                            "email": userInfo["email"] ?? "",
                            "emailVerified": userInfo["verified_email"] ?? false,
                            "name": userInfo["name"] ?? "",
                            "givenName": userInfo["given_name"] ?? "",
                            "familyName": userInfo["family_name"] ?? "",
                            "photoUrl": userInfo["picture"] ?? "",
                            "locale": userInfo["locale"] ?? ""
                        ]
                    ]
                    
                    resolve(result)
                }
            } catch {
                reject("PARSE_ERROR", "Failed to parse user info", error)
            }
        }
        
        task.resume()
    }
    
    private func generateRandomString(length: Int) -> String {
        let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return String((0..<length).map{ _ in letters.randomElement()! })
    }
    
    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc
    override func supportedEvents() -> [String]! {
        return []
    }
}

// MARK: - ASWebAuthenticationPresentationContextProviding

extension GoogleAuthModule: ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        return UIApplication.shared.windows.first { $0.isKeyWindow } ?? ASPresentationAnchor()
    }
}

// MARK: - Dictionary Extension

extension Dictionary {
    func percentEncoded() -> Data? {
        return map { key, value in
            let escapedKey = "\(key)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
            let escapedValue = "\(value)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
            return "\(escapedKey)=\(escapedValue)"
        }
        .joined(separator: "&")
        .data(using: .utf8)
    }
}
